Proto models

kegluneq opened the pull request:
chore(dart/transform): Integrate protoc into gulp build

This change detects if the user has `protoc` available and, if so, uses
it to generate `.pb.dart` files. If not, pre-built files are used
instead.

refactor(dart/transform): Use a protobufs representation during transform

To simplify processing and testing in the future, use protobufs to
represent of `.ng_deps.dart` files rather than always dealing
directly in Dart code.

This update does not actually use the protobuf representation, but this is a
step towards moving all phases to parse and use protobufs rather than Dart
code.

Commit:
chore(dart/transform): Integrate protoc into gulp build

This change detects if the user has `protoc` available and, if so, uses
it to generate `.pb.dart` files. If not, pre-built files are used
instead.

kegluneq commented:
/cc @yjbanov , Yegor, I'd be interested to hear your thoughts on this approach.

The two commits are logically separate from one another. The first consists of the protobufs which will be the representation of our `.ng_deps.dart` files going forward.

The second defines the code that creates those protobufs from the initial `.dart` files and writes them back to `.ng_deps.dart` files when we're done.

The idea would be to have all transformer phases operate on the protobufs and then write to Dart code at the very end. This should speed things up (easier parsing), reduce memory use (fewer ASTs to store), and, especially, reduce the current brittleness in the transformer tests (since we'll no longer be string-compare-ing code).

yjbanov commented:
I think it makes sense, at least as long as we need to deal with multi-phase transformers.

Do we really need to check if the user has `protoc` installed? Why not check in pb files and have a gulp lint that checks that they are up-to-date? This way `protoc` is something only the Angular team needs to deal with.

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/annotation_code.dart:
```diff
@@ -0,0 +1,89 @@
+      // have repeatable output for testing purposes.
+      // Remove this sorting once we are not testing output code directly.
+      var namedParameterMap = new Map.fromIterable(model.namedParameters, key: (e) => e.name, value: (e) => e.value);
+      var sortedKeys = namedParameterMap.keys.toList()..sort();
```
it probably doesn't matter, but it might be a bit cheaper to copy the iterable and sort the entries without creating the map:

```dart
var namedParameters = model.namedParameters.toList();
namedParameters.sort((a, b) => a.name.compareTo(b.name));
for (var param in namedParameters) {
 ...
}
```

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/annotation_code.dart:
```diff
@@ -0,0 +1,89 @@
+  }
+}
+
+class AnnotationWriterMixin {
```
+ short dart doc

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/annotation_code.dart:
```diff
@@ -0,0 +1,89 @@
+}
+
+class AnnotationWriterMixin {
+  StringBuffer buffer;
```
maybe make this an abstract getter and mark the class abstract?

(look like this is a field that you expect to be defined in the class that mixes this Mixin, correct?)

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/constify.dart:
```diff
@@ -0,0 +1,62 @@
+    if (hasArguments) {
+      var args = node.arguments.arguments;
+      writer.print('(');
+      for (var i = 0, iLen = args.length; i < iLen; ++i) {
```
(btw, I think the vm knows about these patterns and may optimize it if you remove iLen, but check for .length on every iteration)

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/import_export_code.dart:
```diff
@@ -0,0 +1,95 @@
+
+/// Class to be used as a mixin to allow writing [ImportModel]s to Dart code.
+class ImportWriterMixin {
+  StringBuffer buffer;
```
same here and in the other mixins below (mark abstract and make buffer a getter)

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/injectable_code.dart:
```diff
@@ -0,0 +1,185 @@
+
+/// Visitor responsible for parsing [ClassDeclaration]s into [InjectableModel]s.
+class InjectableVisitor extends Object
+    with RecursiveAstVisitor<InjectableModel> {
```
could this be written as: `class InjectableVisitor extends RecursiveAstVisitor<InjectableModel>`?

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/injectable_code.dart:
```diff
@@ -0,0 +1,185 @@
+      ctorName = ctorName == null
+          ? 'the unnamed constructor'
+          : 'constructor "${ctorName}"';
+      logger.warning('Found ${numCtorsFound} ctors for class ${node.name},'
```
I wonder if this warning would be too common - maybe we can skip the warning when we find the unnamed constructor, but only print it when we had to pick a named constructor?

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/injectable_code.dart:
```diff
@@ -0,0 +1,185 @@
+}
+
+class InjectableWriterMixin {
+  StringBuffer buffer;
```
same here (get buffer)

sigmundch replied:
(similarly the 2 fields below)

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/injectable_code.dart:
```diff
@@ -0,0 +1,185 @@
+
+class InjectableWriterMixin {
+  StringBuffer buffer;
+  AnnotationWriterMixin annotationWriter;
```
it's unfortunate that Dart doesn't allow you to do mixin composition here :(.

We could however play the trick of declaring abstract members or abstract interfaces. For example:

```dart
abstract class InjectableWriterMixin implements AnnotationWriterMixin, ParameterWriterMixin {
  StringBuffer get buffer;
  void _writeList() { ...
     writeAnnotationModel(...);
  }
}
```

It looks strange at first, but it gives you the right set of static checks: inside this class you can use anything declared in those interfaces, and you are forcing however mixes this mixin to also include an implementation for writeAnnotationModel and writeParameterModel*. The class that mixes this mixin and AnnotationWriterMixin and ParameterWriterMixin will just work.

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/common/code/parameter_code.dart:
```diff
@@ -0,0 +1,127 @@
+
+/// Visitor responsible for parsing [FormalParameter]s into
+/// [ParameterModel]s.
+class ParameterVisitor extends Object with SimpleAstVisitor<ParameterModel> {
```
`extends Object with Simple...` => `extends Simple...` ?

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/test/transform/integration/two_deps_files/expected/bar.ng_deps.dart:
```diff
@@ -18,8 +18,8 @@ void initReflector() {
-          const [prefix.MyContext],
-          const [String]
+          const [prefix.MyContext,],
+          const [String,]
```
might not be relevant, but looks like the const printer is including a trailing comma

kegluneq replied:
Done

sigmundch commented on modules_dart/transform/lib/src/transform/directive_processor/inliner.dart:
```diff
@@ -0,0 +1,132 @@
+        var rawTemplateUrl = _getTemplateUrlValue(annotation);
+        if (rawTemplateUrl != null) {
+          if (_hasTemplateValue(annotation)) {
+            logger.warning('Both template url and template are specified. '
```
We might want to remove this warning. I couldn't find it in the code before the change, so I assume it is new? This relates to a change I did in a transformer about a month ago: I was fixing how we resolve urls that are seen inside style-tags after we did inlining (e.g. we inlined the HTML but there was a `@import` inside the CSS of a style tag). To fix that, we decided to include both the template and templateUrl after inlining.

kegluneq replied:
This is in the code that actually does that inline-ing -- it should only occur if the user has specified both a `template` and a `templateUrl`, which doesn't really make sense.

sigmundch commented:
Very nice!

I agree with Yegor in that we can restrict our dependency on protoc and simply use it in the angular repo to update the .pb.dart files and to validate that the submitted .pb.dart files are up-to-date.

kegluneq commented:
Updated the `.pb.dart` generation code & added a lint step to `build/packages.dart`.

That could probably use some guidance from somebody more comfortable with `gulp` than I am, but what do you think @sigmundch and @yjbanov about the general thrust?

I'll take a look at the code comments now.

kegluneq commented:
I think this is ready - can you PTAL at the changes?

@yjbanov, do you have a recommendation for who I should speak to about cleaning up the `gulp` implementation?

yjbanov commented:
@IgorMinar is most experience with our build setup.

kegluneq commented:
`ReflectionInfoModel` works for me - updated.

Any other objections?

sigmundch commented:
lgtm

IgorMinar commented on gulpfile.js:
```diff
@@ -1231,8 +1232,117 @@ gulp.task('clean', ['build/clean.tools', 'build/clean.js', 'build/clean.dart', '
+  // Dart sdk is not available.
+  var protoc = require('./tools/build/protoc').detect();
+  if (!protoc) {
+    console.log('INFO: Could not detect protoc - failed to rebuild Dart proto code.');
```
this should be an error and not just a console.log. no?

kegluneq replied:
True, will fix.

IgorMinar commented on gulpfile.js:
```diff
@@ -833,7 +833,7 @@ gulp.task('build/pure-packages.dart', function() {
     .src([
       'modules_dart/transform/**/*',
-      '!modules_dart/transform/pubspec.yaml'
+      '!modules_dart/transform/**/*.proto'
```
is this right? looks like an odd change.

kegluneq replied:
The original line looks obsolete, there is no `pubspec.yaml` in `modules_dart/transform`.

I can leave it in if you prefer though.

The new line is accurate -- since we're shipping the `.pb.dart` files we don't need the `.proto`s

IgorMinar commented:
I'm missing a ton of context but it looks weird that we generate some code on the fly and other code we generate once and store the artifacts in the repo. Is there a good reason for this?

I also see that the new tasks are not being used by anything, but I assume that that will still come. is that right?

Also the gulp file is becoming a giant mess, could you externalize all the transformer task into a separate file and include it in the main gulp file?

kegluneq commented:
@IgorMinar , some context below. What would be your preference?

Generating the `.pb.dart` files only needs to happen when the `.proto` files change, and generating them requires `protoc`, so I'm hoping this is something that only the Angular team will need to deal with.

That being said, if the protos do change, we want to know about it, so that is the rationale for the `lint_protos.dart` target.

The `gen_protos.dart` target is not included in any of the build pipeline because it should just be run on demand, whenever the protos change.

Alternatives I considered are:
* Have a script that generates the protos in `scripts/build` or somewhere similar. This seems worse to me than having this in `glupfile`.
* Create a wiki page explaining how to regenerate the `.pb.dart` files. This has more overhead than just calling a `gulp` target, but is less magical and doesn't clutter the project.

kegluneq commented:
@IgorMinar , I updated `gulpfile.js` with the changes we discussed, ptal

IgorMinar commented:
Lgtm for the gulp changes.

On Mon, Sep 14, 2015, 4:45 PM Timothy Blasi <notifications@github.com>
wrote:

> @IgorMinar <https://github.com/IgorMinar> , I updated gulpfile.js with
> the changes we discussed, ptal
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular/pull/4105#issuecomment-140234644>.
>

Commit:
refactor(dart/transform): Use a protobufs representation during transform

To simplify processing and testing in the future, use protobufs to
represent of `.ng_deps.dart` files rather than always dealing
directly in Dart code.

This update does not actually use the protobuf representation, but this
is a step towards moving all phases to parse and use protobufs rather than
Dart code.
